### ISSUE TITLE
authorize: allow arbitrary claims in assertion / identity headers

### ISSUE THREAD

**Issue by desimone** [pomerium/pomerium#1855]
**Is your feature request related to a problem? Please describe.**

Today, it possible to get a [user's identity](https://www.pomerium.com/docs/topics/getting-users-identity.html#verification) either from a signed attestation header (`X-Pomerium-Jwt-Assertion`) or via a subset of optional unisgned [identity headers](https://www.pomerium.com/reference/#jwt-claim-headers) (`X-Pomerium-Claim-Email` / `X-Pomerium-Claim-Group`).

However, is not possible to add identity details in a flexible way to support systems that may expect an unsigned identity header with a different header key (commonly, `x-forwarded-user`) . 

Also, upstream applications may desire more flexibility in what identity details (both to enrich, and prevent leaking unwanted details) are included in the signed assertion jwt header. 

**Describe the solution you'd like**

For unsigned identity headers:
- It should be possible to define the desired header key (e.g. `x-forwarded-user`) for a given claim value. 
- It should be possible to assign an arbitrary datastore value (possibly using rego syntax? here I'm less sure about the implementation) to a given header key. 

For the signed assertion header:

- It should be possible to define an arbitrary set of identity claims to the assertion JWT. 

To prevent leaking data, both should be configurable on a per-policy (vs route?) basis. 

**Describe alternatives you've considered**

I suppose we could support arbitrary header rewriting as a workaround, but that feels janky, and does not handle the assertion use case. 

```
   RewriteEngine On
   RewriteCond %{LA-U:REMOTE_USER} (.+)
   RewriteRule . - [E=RU:%1,NS]
   RequestHeader set X-Forwarded-User "%{RU}e" env=RU
```

**Explain any additional use-cases**

If there are any use-cases that would help us understand the use/need/value please share them as they can help us decide on acceptance and prioritization.

**Additional context**

- [Jenkins](https://plugins.jenkins.io/reverse-proxy-auth-plugin/) wanting `X-Forwarded-User` `X-Forwarded-Groups`
- [zammad](https://docs.zammad.org/en/latest/appendix/single-sign-on.html#conceptual-overview) wanting `X-Forwarded-User`

**Comment by travisgroth** [comment]
After discussion - for unsigned identity headers we should only support the claims as we do today.  Exposing the databroker values arbitrarily will need more thought and documentation.

1. Allow customization of the unsigned identity claim header names
1. Allow adding arbitrary claims to the assertion header
1. Support these on a per-route basis, since the needs likely vary by route
1. Should we support a configurable list separator?  

Given the last requirement it seems like this is a net-new option field and could be a breaking change, assuming we remove `jwt_claims_headers` from being in the top level.

Possible syntax for (1):

```yaml
from: https://foo
to: https://bar
jwt_claims_headers:
  x-custom-header-sub: sub
  x-custom-header-groups: groups
```

The resulting headers should be:
```
X-Custom-Header-Sub: okta/00usi7mc8XC8SwFxT4x6
X-Custom-Header-Groups: group1,group2,group3
```

@desimone not sure if we should rename this option from `jwt_claims_headers` if it is a breaking change.

**Comment by calebdoxsey** [comment]
I implemented the claim headers. I'm not sure how to implement adding data to the assertion header.

**Comment by rawtaz** [comment]
Here's a potentially very stupid question, but nonetheless: Shouldn't it be possible to define what we want the signed attestation header to be?

For example, at https://jwt.io/introduction/ it says under "How do JSON Web Tokens work?" that "Whenever the user wants to access a protected route or resource, the user agent should send the JWT, typically in the Authorization header using the Bearer schema. The content of the header should look like the following:"

~~~
Authorization: Bearer <token>
~~~

I see `Authorization` being the recommended name of the header for this in other places too. So, in case there's software that is looking at this header (and worst case isn't configurable to look at another header name), wouldn't it be good if we could at least configure the name of the header, as it's currently not the seemingly more conventional `Authorization` one?

Apologies if I'm completely out in the dark here.

**Comment by desimone** [comment]
Passing arbitrary claims and setting those claims to specific keys is now supported (by @calebdoxsey ).

Related:

- https://github.com/pomerium/pomerium/pull/2102
- https://github.com/pomerium/pomerium/pull/1906 
- https://github.com/pomerium/pomerium-console/issues/752

**Comment by desimone** [comment]
There was a regression in `v0.13` and `v0.12` but this should be fixed again in `v0.14`. 

https://pomerium-io.slack.com/archives/CK92MUAES/p1620051083096200?thread_ts=1619723701.078300&cid=CK92MUAES
